
/*
#### Compute Gene Stoichiometry and output a summary table from a _hardcoded_ file path.

**Execute this file as `hyphy GS-01.md`**

First, make sure we require that the user currently has a recent version of HyPhy.

```
*/
    RequireVersion ("2.4.0");
/*
```

Here, we will use some of the the libv3 functions available in HyPhy. They will be used later in our script, and they are imported using the LoadFunctionLibrary command by specifying the location of the target batch file withing the HyPhy Resources folder.


```
*/
LoadFunctionLibrary ("libv3/tasks/alignments.bf");
LoadFunctionLibrary("libv3/IOFunctions.bf");
LoadFunctionLibrary("libv3/UtilityFunctions.bf");
/*
```

We will rely on the known chemical structure for the nucleotides to inform us about their respective atomic consituents. This will be used later in the script, after we calculate the count of nucleotides (A, T, C, G) in order to calculate the number of Carbon, Nitrogen, Oxygen, and Hydrogen atoms within each gene sequence.

```
*/

A = {"Carbon": 5, "Nitrogen": 5, "Oxygen": 0, "Hydrogen": 5};
T = {"Carbon": 5, "Nitrogen": 2, "Oxygen": 2, "Hydrogen": 6};
C = {"Carbon": 4, "Nitrogen": 2, "Oxygen": 1, "Hydrogen": 6};
G = {"Carbon": 4, "Nitrogen": 6, "Oxygen": 1, "Hydrogen": 5};

/*
```
1. Read the alignment from a hardcoded file (relative to the current working directory) using the built-in `ReadDataFile` command.
```
*/
fprintf (stdout, "# Reading fasta file 'ncov.fas' from current directory \n");
DataSet alignment = ReadDataFile ('ncov.fas');
fprintf (stdout, alignment, "\n");
/*
```

The number of sequences in `alignment` are stored in the `alignment.species` variable.

Next, we will set up some of the output ("Summary") table configurations, and column headers. This is controlled through 'output.table_output_options' and 'output.report'. After this is configured, the columns headers are printed to the console.


```
*/
output.table_output_options = {
    utility.getGlobalValue("terms.table_options.header"): TRUE,
    utility.getGlobalValue("terms.table_options.minimum_column_width"): 10,
    utility.getGlobalValue("terms.table_options.align"): "center",
    utility.getGlobalValue("terms.table_options.column_widths"): {
        "0": 22,
        "1": 30,
        "2": 22,
        "3": 22,
        "4": 22,
        "5": 22,
        "6": 22
    }
};

output.report = {
    {
        "Sequence ID", "Sequence Length (ungapped)", "Total Carbon Atoms", "Total Hydrogen Atoms", "Total Oxygen Atoms", "Total Nitrogen Atoms", "Molecular Weight"
    }
};

fprintf (stdout, "\n### SUMMARY ###\n");

fprintf(stdout, "\n",  io.FormatTableRow(output.report, output.table_output_options));
/*
```

1. We set up a C-style loop to iterate over the sequences, and dictionary variables (`nucleotides_per_seq`) to keep track of how many nucleotide characters we found per sequence
2. During each iteration we retrieve sequence strings using the built-in `GetDataInfo` procedure and get the sequence ID from the built-in `GetString` procedure
3. For each sequence, we iterate over each character in the string (`Abs(string)` is the length of the string) and add a count to the dictionary.
4. We calculate the ungapped sequence length from the sum of 'A', 'T', 'C', 'G'
5. We calculate the number of Carbon, Nitrogen, Oxygen, and Hydrogen atoms in each sequences.
5. Finally, we print the resulting data table to the console


```
*/

for (sequence_index = 0; sequence_index < alignment.species; sequence_index+=1) {
  GetDataInfo (ith_sequence, alignment, sequence_index);
  GetString (seq_name, alignment, sequence_index);
  
  nucleotides_per_seq = {};

  for (character_index = 0; character_index < Abs (ith_sequence); character_index += 1) {
    nucleotides_per_seq [ith_sequence[character_index]] += 1;
  } // end inner for
  
  seq_length = nucleotides_per_seq['A'] + nucleotides_per_seq['C'] + nucleotides_per_seq['G'] + nucleotides_per_seq['T'];
  
  // Print row to table. 
  // Molecular weight calcuation is not yet implemented.
  fprintf(stdout, io.FormatTableRow(
      {
       {
       seq_name,
       seq_length,
       (nucleotides_per_seq['T'] * T["Carbon"]) + (nucleotides_per_seq['C'] * C["Carbon"]) + (nucleotides_per_seq['G'] * G["Carbon"]) + (nucleotides_per_seq['A'] * A["Carbon"]),
       (nucleotides_per_seq['T'] * T["Hydrogen"]) + (nucleotides_per_seq['C'] * C["Hydrogen"]) + (nucleotides_per_seq['G'] * G["Hydrogen"]) + (nucleotides_per_seq['A'] * A["Hydrogen"]),
       (nucleotides_per_seq['T'] * T["Oxygen"]) + (nucleotides_per_seq['C'] * C["Oxygen"]) + (nucleotides_per_seq['G'] * G["Oxygen"]) + (nucleotides_per_seq['A'] * A["Oxygen"]),
       (nucleotides_per_seq['T'] * T["Nitrogen"]) + (nucleotides_per_seq['C'] * C["Nitrogen"]) + (nucleotides_per_seq['G'] * G["Nitrogen"]) + (nucleotides_per_seq['A'] * A["Nitrogen"]),
       "N/A"
       }
      },
      output.table_output_options)
  );
  
} // End outer for

fprintf (stdout, "\n");





/*
```
*/




// ##############################
//   End of file
// ##############################


