/*
#### Compute GC content and CpG deficiency from a _hardcoded_ file path using more efficient built-in functions

**Execute this file as `hyphy CG-02.md`**

> See [the first part of this tutorial](CG-01.md) for some background

Here we will use internal procedures `HarvestFrequencies` and `DataSetFilter`

1. Read the alignment from a hardcoded file (relative to the current working directory) using the built-in `ReadDataFile` command.

```
*/
      DataSet alignment = ReadDataFile ('ncov.fas');
/*
```
Use the `HarvestFrequencies` procedure to estimate the frequencies of nuclotides in the dataset. The procedure takes five positional arguments

`HarvestFrequencies (storage, source, atom, unit, position-aware)`

* `storage` is a variable that will be populated (as a Matrix) with estimated frequencies.
* `source` is a `DataSet` (or `DataSetFilter`, see below)
* `atom` (integer > 0) counts how many alignment columns are in an evolutionary unit (e.g. for codon alignments this is 3)
* `unit` (integer > 0 and â‰¤ `atom`, divides `atom`) tells HyPhy how to count within the evolutionary unit
* `position-aware` (boolean, only matters if `unit` < `atom`) creates position-specific alignment frequencies 

First, we compute the overall CG content. `overall_frequencies` is going to receive a 4x1 matrix, with `A`, `C`, `G`, `T` frequencies, so the sum of the second and the third entries is the desired quantity. Note that `HarvestFrequencies` will automatically handle ambiguous nucleotides, including `?` added via padding by `ReadDataFile` (which assumes that aligned sequences are supplied)

```
*/

HarvestFrequencies (overall_frequencies, alignment, 1, 1, 1);
overall_cg = overall_frequencies[1] + overall_frequencies[2]; 

fprintf (stdout, "Overall CG content is ", Format (overall_cg * 100, 8, 3), "%\n");

/*
```

Next we need to evaluate the frequency of all di-nucleotides to estimate Cp deficiency. This is a little tricky, but shows some useful HyPhy features. We are going to use 
`HarvestFrequencies (dinucleotide_frequencies, alignment, 2, 2, FALSE);` to estimate the frequencies of all dinucleotides in the data, but we need **overlapping** counts, e.g. in the sequence `GATTAC`, we need to count `GA`, `AT`, `TT`, `TA`, `AC`,`CA`, whereas `HarvestFrequencies` will only count non-overlapping dinucleotides, i.e. `GA`, `TT`, `AC`. A simple solution is to call `HarvestFrequencies` twice -- once of the original data, and once on the data starting with the second nucletide position, which is where the `DataSetFilter` command comes into play

1. Estimate non-overlapping di-nucleotide frequencies (16x1 matrix, the entry for CG is in obtained from it's alphanumerical order, 7-th, or 6 in 0-based indexing)
2. Create a filter of the data, where the first column of the alignment is skipped.
3. Estimate non-overlapping di-nucleotide frequencies for the shifted data
4. Average the two

The slight difference between the results here and in the first tutorial are because `HarvestFrequencies` will count `?` as contribting a 1/4 count to each of the nucleotides.

```
*/

HarvestFrequencies (dinucleotide_frequencies, alignment, 2, 2, FALSE);
DataSetFilter shifted = CreateFilter (alignment, 1, siteIndex > 0);
HarvestFrequencies (dinucleotide_frequencies_shifted, shifted, 2, 2, FALSE);
CpG = 0.5 * (dinucleotide_frequencies[6] + dinucleotide_frequencies_shifted[6]);

fprintf (stdout, "CpG deficiency is ", Format (CpG/(overall_frequencies[1]* overall_frequencies[2]), 8, 3), "\n");

/*
```
*/
